### DISCUSSION TITLE
strange table behaviour

### DISCUSSION THREAD

**Discussion by more-urgent-jest** [mganss/HtmlSanitizer#561] [Q&A]
Hi, I am trying to restrict the allowed tags by clearing the set and adding a restricted list like so:
``` csharp
        var sanitizer = new HtmlSanitizer();

        sanitizer.AllowedTags.Clear();
        sanitizer.AllowedTags.UnionWith(["ul", "li", "table", "tr", "td", "strong"]);
        sanitizer.AllowedAttributes.Clear();
        sanitizer.AllowedAttributes.Add("class");
        sanitizer.AllowedAtRules.Clear();
        sanitizer.AllowedClasses.Clear();
        sanitizer.AllowedClasses.Add("right");
        sanitizer.AllowedCssProperties.Clear();
        sanitizer.AllowedSchemes.Clear();

        services.AddSingleton<IHtmlSanitizer>(sanitizer);
```
but finding that it strips tr tags and their children and content, but just stripping td tags and leaving their content.

this:
"\<ul>\<li>the item in this list containing this \<tr> \</tr> \<td>should stay\</td> \<table class="right"> \</table> as it is\</li>\</ul>"
becomes:
"\<ul>\<li>the item in this list containing this   should stay \<table class="right"> \</table> as it is\</li>\</ul>"

this:
"\<ul>\<li>the item in this list containing this \<tr> \<td>should stay\</td> \</tr> \<table class="right"> \</table> as it is\</li>\</ul>"
becomes:
"\<ul>\<li>the item in this list containing this  should stay  \<table class="right"> \</table> as it is\</li>\</ul>"

this:
"\<ul>\<li>the item in this list containing this \<table class="right"> \<tr> \<td>should stay\</td> \</tr> \</table> as it is\</li>\</ul>"
becomes:
"\<ul>\<li>the item in this list containing this \<table class="right"> \</table> as it is\</li>\</ul>"

this:
"\<ul>\<li>the item in this list containing this \<table>\<tr class="right">\<td>should stay\</td>\</tr>\</table> as it is\</li>\</ul>"
becomes:
"\<ul>\<li>the item in this list containing this \<table>\</table> as it is\</li>\</ul>"

I would expect all of these to not be changed.

am I not understanding how this works?

**Comment by more-urgent-jest** [comment]
seems to be happening, because the parser inserts \<tbody>, but with it in there I am getting this even stranger behaviour:

this: 
"\<ul>\<li>the item in this list containing this \<table>\<tbody>\<tr class="right">\<td>should stay\</td>\</tr>\<tbody>\</table> as it is\</li>\</ul>"
becomes:
"\<ul>\<li>the item in this list containing this \<table>\<tbody>\<tr class="right">\<td>should stay\</td>\</tr>\</tbody>\<tbody>\</tbody>\</table> as it is\</li>\</ul>"

which seems especially odd as tr is not required to be in a tbody, but can be contained directly within a table, as specified in the table content model:

> In this order: optionally a [caption](https://html.spec.whatwg.org/multipage/tables.html#the-caption-element) element, followed by zero or more [colgroup](https://html.spec.whatwg.org/multipage/tables.html#the-colgroup-element) elements, followed optionally by a [thead](https://html.spec.whatwg.org/multipage/tables.html#the-thead-element) element, followed by either zero or more [tbody](https://html.spec.whatwg.org/multipage/tables.html#the-tbody-element) elements or one or more [tr](https://html.spec.whatwg.org/multipage/tables.html#the-tr-element) elements, followed optionally by a [tfoot](https://html.spec.whatwg.org/multipage/tables.html#the-tfoot-element) element, optionally intermixed with one or more [script-supporting elements](https://html.spec.whatwg.org/multipage/dom.html#script-supporting-elements-2).

[the table element](https://html.spec.whatwg.org/multipage/tables.html#the-table-element)

and the tr contexts:

> As a child of a [thead](https://html.spec.whatwg.org/multipage/tables.html#the-thead-element) element.
As a child of a [tbody](https://html.spec.whatwg.org/multipage/tables.html#the-tbody-element) element.
As a child of a [tfoot](https://html.spec.whatwg.org/multipage/tables.html#the-tfoot-element) element.
As a child of a [table](https://html.spec.whatwg.org/multipage/tables.html#the-table-element) element, after any [caption](https://html.spec.whatwg.org/multipage/tables.html#the-caption-element), [colgroup](https://html.spec.whatwg.org/multipage/tables.html#the-colgroup-element), and [thead](https://html.spec.whatwg.org/multipage/tables.html#the-thead-element) elements, but only if there are no [tbody](https://html.spec.whatwg.org/multipage/tables.html#the-tbody-element) elements that are children of the [table](https://html.spec.whatwg.org/multipage/tables.html#the-table-element) element.

[the tr element](https://html.spec.whatwg.org/multipage/tables.html#the-tr-element)

**Comment by more-urgent-jest** [comment]
actually, I had a second opening tbody instead of a closing one, so that behaviour isn't that strange after all, but it is odd that tbody is required when it shouldn't be.

I would have hoped that since AngleSharp, if indeed that is causing this, claims to be standards compliant that it wouldn't exhibit such eccentricities.

**Reply by tiesont** [reply]
I assume AngleSharp is following browser behavior, which is to insert a `<tbody>` (around all of the content) into the table if one is not present, which then gets stripped out if tbody isn't in the allowed tags list.

Beyond that, an MVCE would be helpful to verify what you're seeing. 

Also, if you're trying to use code fences (```) in your code above, then the opening and closing fences need to be on their own lines (Github is currently parsing your text with inline code blocks, which are just a single backtick on either side of the code). Just FYI.
